core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java

```diff
@@ -0,0 +1,372 @@
+                middle.set(middle.size() - 1, new Any(last.min(), false));
+            }
+        }
+
+        NFA nfa = makeNfa(middle);
+        DFA dfa = nfa.toDFA();
+
+        int[] transitions = new int[dfa.states().size() * 256];
+        boolean[] accept = new boolean[dfa.states().size()];
+
+        for (DFA.State state : dfa.states()) {
+            for (DFA.Transition transition : dfa.transitions(state)) {
+                transitions[state.id() * 256 + transition.value()] = transition.target().id() * 256;
+            }
+
+            if (state.accept()) {
+                accept[state.id()] = true;
+            }
+        }
+
+        return new LikeMatcher(
+                pattern,
+                escape,
+                minSize,
+                unbounded ? OptionalInt.empty() : OptionalInt.of(maxSize),
+                prefix,
+                suffix,
+                new DenseDfaMatcher(transitions, dfa.start().id(), accept, 0, exact));
+    }
+
+    public boolean match(byte[] input)
+    {
+        return match(input, 0, input.length);
+    }
+
+    public boolean match(byte[] input, int offset, int length)
+    {
+        if (length < minSize) {
+            return false;
+        }
+
+        if (maxSize.isPresent() && length > maxSize.getAsInt()) {
+            return false;
+        }
+
+        if (!match(prefix, input, offset)) {
+            return false;
+        }
+
+        if (!match(suffix, input, offset + length - suffix.length)) {
+            return false;
+        }
+
+        return matcher.match(input, offset + prefix.length, length - suffix.length - prefix.length);
+    }
+
+    /**
+     * Ends as soon as it reaches an accept state. This is useful for
+     * patterns that end in %
+     */
+    private boolean match(byte[] pattern, byte[] input, int offset)
+    {
+        for (int i = 0; i < pattern.length; i++) {
+            if (pattern[i] != input[offset + i]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static List<Pattern> parse(String pattern, Optional<Character> escape)
+    {
+        List<Pattern> result = new ArrayList<>();
+
+        StringBuilder literal = new StringBuilder();
+        boolean inEscape = false;
+        for (int i = 0; i < pattern.length(); i++) {
+            char character = pattern.charAt(i);
+
+            if (inEscape) {
+                if (character != '%' && character != '_' && character != escape.get()) {
+                    throw new IllegalArgumentException("Escape character must be followed by '%', '_' or the escape character itself");
+                }
+                literal.append(character);
+                inEscape = false;
+            }
+            else if (escape.isPresent() && character == escape.get()) {
+                inEscape = true;
+            }
+            else if (character == '%' || character == '_') {
```

Comment:
a bit tangential, does sql spec allow % and _ as escapes too?

Reply:
Yes, it allows any character as an escape.